File: windows/tests/ConnectorControl.Core.Tests/TestSupport/FakeAutostart.cs
using ConnectorControl.Core.Services;

namespace ConnectorControl.Core.Tests.TestSupport;

public sealed class FakeAutostart : IAutostart
{
    public bool Enabled { get; set; }
    /// <summary>When set, SetEnabled throws InvalidOperationException with this message.</summary>
    public string? FailWith { get; set; }
    public int SetCalls { get; private set; }

    public bool IsEnabled => Enabled;

    public void SetEnabled(bool enabled)
    {
        SetCalls++;
        if (FailWith is not null)
        {
            throw new InvalidOperationException(FailWith);
        }
        Enabled = enabled;
    }
}

File: Tests/ConnectorControlStateTests/TestSupport/FakeAutostart.swift
import Foundation
@testable import ConnectorControlState

struct FakeAutostartError: LocalizedError {
    let message: String
    var errorDescription: String? { message }
}

final class FakeAutostart: Autostart {
    var enabled = false
    var requiresApproval = false
    /// When set, setEnabled throws an error whose localizedDescription is this text.
    var failWith: String?
    private(set) var setCalls = 0

    var isEnabled: Bool { enabled }

    func setEnabled(_ on: Bool) throws {
        setCalls += 1
        if let failWith { throw FakeAutostartError(message: failWith) }
        enabled = on
    }
}
